### DISCUSSION TITLE
Pre-initialize project import (warm LS)?

### DISCUSSION THREAD

**Discussion by guw** [redhat-developer/vscode-java#3489] [Q&A]
Did anyone experiment with pre-initializing the project import?

Our import takes a couple minutes and we are looking for a way to pre-initialize the workspace in a background (headless) process. Thus, when the user brings up VS Code, the Java extension would activate much faster as it already has a workspace populated with Java projects.

**Comment by rgrunber** [comment]
Have you tried with with setting `"java.autobuild.enabled": "false"` ? I think the part that often takes longest in the project import is the build (which by default happens automatically).

I would think that if the contents of the workspace storage are in place, that ultimately reduces the import time also. Those contents are stored in VS Code's data directory, under `User/workspaceStorage/${projectFolderID}`, where `${projectFolderID}` (at least on Linux) [is the md5 checksum of the folder's inode number, appended to its absolute path](https://github.com/microsoft/vscode/blob/main/src/vs/platform/workspaces/node/workspaces.ts#L49-L78) (eg. `${folder_path}${inode}`). You can also just check the `workspace.json` file inside each folder ID for the absolute folder path associated with it. Under `redhat.java/jdt_ws`, you'll find that it's basically the workspace's state location (`.metadata/`) that ends up having most of the important data. If that directory were shared, it might be enough to reduce the import time, but I haven't really played with this idea.

There has been work done to reduce indexing through things like `java.sharedIndexes.enabled`, but that doesn't drastically improve the time so it happens in the background.

**Update:** Another thing is that the first time one launches a new version of vscode-java, there are a lot of files that get cached into `User/globalStorage/redhat.java/${version}/config_*`. All of the jars that are nested into OSGi bundles need to get cached there to be referenced (eg. think `Bundle-ClassPath`), as well as any wiring instructions that may have taken time to compute on the initial startup.

**Reply by guw** [reply]
No. I do not want to disable auto building. I really want to "pre-import" projects. Our import is specifically expensive.
